File: src/kinematic_bicycle/kinematic_bicycle/mpc/constraints.py
"""Constraint definitions for MPC

Keeps constraint logic organized and reusable
"""

import numpy as np
from .config import MPCConfig


class ConstraintSet:
    """Container for MPC constraints
    
    Handles box constraints and custom constraint logic
    """
    
    def __init__(self, config: MPCConfig):
        """Initialize constraints from config
        
        Args:
            config: MPC configuration
        """
        self.config = config
    
    def get_input_bounds(self) -> tuple:
        """Get control input bounds
        
        Returns:
            (lower_bound, upper_bound): Shape (horizon, 2)
        """
        horizon = self.config.horizon
        
        # Velocity bounds [v_min, v_max]
        v_bounds = np.array([self.config.v_min, self.config.v_max])
        
        # Steering rate bounds [-delta_dot_max, delta_dot_max]
        delta_dot_bounds = np.array([-self.config.delta_dot_max, self.config.delta_dot_max])
        
        lower = np.zeros((horizon, 2))
        upper = np.zeros((horizon, 2))
        
        lower[:, 0] = self.config.v_min
        upper[:, 0] = self.config.v_max
        
        lower[:, 1] = -self.config.delta_dot_max
        upper[:, 1] = self.config.delta_dot_max
        
        return lower, upper
    
    def get_state_bounds(self) -> tuple:
        """Get state bounds
        
        For now, mainly steering angle constraint
        
        Returns:
            (lower_bound, upper_bound): Shape (horizon+1, 4)
        """
        horizon = self.config.prediction_size
        
        lower = np.full((horizon, 4), -np.inf)
        upper = np.full((horizon, 4), np.inf)
        
        # Steering angle constraint: -delta_max <= delta <= delta_max
        lower[:, 3] = -self.config.delta_max
        upper[:, 3] = self.config.delta_max
        
        return lower, upper
    
    def check_feasibility(self, state: np.ndarray, control: np.ndarray) -> bool:
        """Check if state-control pair satisfies constraints
        
        Args:
            state: [x, y, theta, delta]
            control: [v, delta_dot]
        
        Returns:
            bool: True if feasible
        """
        # Steering angle constraint
        if not (-self.config.delta_max <= state[3] <= self.config.delta_max):
            return False
        
        # Velocity constraint
        if not (self.config.v_min <= control[0] <= self.config.v_max):
            return False
        
        # Steering rate constraint
        if not (-self.config.delta_dot_max <= control[1] <= self.config.delta_dot_max):
            return False
        
        return True

File: src/mpc_controller/src/Constraints/constraints.cpp
#include "constraints.h"
#include <cmath>

namespace mpc_controller {

///////////////////////////////////////////////////////////////////////////////
// Constructor
///////////////////////////////////////////////////////////////////////////////
ConstraintSet::ConstraintSet(const Params& params) : params_(params) {}

///////////////////////////////////////////////////////////////////////////////
// STATE BOUNDS [x, y, theta, delta] - Explicit & Clean
///////////////////////////////////////////////////////////////////////////////
Eigen::Vector4d ConstraintSet::getStateLowerBounds() const {
    // Return lower bounds for states: [x, y, theta, delta]
    return Eigen::Vector4d(
        params_.x_min,           // 0: x position lower bound
        params_.y_min,           // 1: y position lower bound
        params_.theta_min,       // 2: heading angle lower bound
        params_.delta_min        // 3: steering angle lower bound
    );
}

Eigen::Vector4d ConstraintSet::getStateUpperBounds() const {
    // Return upper bounds for states: [x, y, theta, delta]
    return Eigen::Vector4d(
        params_.x_max,           // 0: x position upper bound
        params_.y_max,           // 1: y position upper bound
        params_.theta_max,       // 2: heading angle upper bound
        params_.delta_max        // 3: steering angle upper bound
    );
}

///////////////////////////////////////////////////////////////////////////////
// INPUT BOUNDS [v, delta_dot] - Explicit & Clean
///////////////////////////////////////////////////////////////////////////////
Eigen::Vector2d ConstraintSet::getInputLowerBounds() const {
    // Return lower bounds for inputs: [v, delta_dot]
    return Eigen::Vector2d(
        params_.v_min,           // 0: velocity lower bound
        -params_.delta_dot_max   // 1: steering rate lower bound (symmetric)
    );
}

Eigen::Vector2d ConstraintSet::getInputUpperBounds() const {
    // Return upper bounds for inputs: [v, delta_dot]
    return Eigen::Vector2d(
        params_.v_max,           // 0: velocity upper bound
        params_.delta_dot_max    // 1: steering rate upper bound (symmetric)
    );
}

///////////////////////////////////////////////////////////////////////////////
// MATRIX VERSIONS FOR MPC HORIZON (Convenience methods)
///////////////////////////////////////////////////////////////////////////////
std::pair<Eigen::MatrixXd, Eigen::MatrixXd> ConstraintSet::getInputBounds() const {
    int horizon = params_.horizon;
    
    Eigen::MatrixXd lower = Eigen::MatrixXd::Zero(horizon, 2);
    Eigen::MatrixXd upper = Eigen::MatrixXd::Zero(horizon, 2);
    
    // Apply same bounds for all steps using explicit vector method
    Eigen::Vector2d input_lower = getInputLowerBounds();
    Eigen::Vector2d input_upper = getInputUpperBounds();
    
    for (int i = 0; i < horizon; ++i) {
        lower.row(i) = input_lower.transpose();
        upper.row(i) = input_upper.transpose();
    }
    
    return {lower, upper};
}

std::pair<Eigen::MatrixXd, Eigen::MatrixXd> ConstraintSet::getStateBounds() const {
    int horizon = params_.getPredictionSize();
    
    Eigen::MatrixXd lower = Eigen::MatrixXd::Zero(horizon, 4);
    Eigen::MatrixXd upper = Eigen::MatrixXd::Zero(horizon, 4);
    
    // Apply same bounds for all steps using explicit vector method
    Eigen::Vector4d state_lower = getStateLowerBounds();
    Eigen::Vector4d state_upper = getStateUpperBounds();
    
    for (int i = 0; i < horizon; ++i) {
        lower.row(i) = state_lower.transpose();
        upper.row(i) = state_upper.transpose();
    }
    
    return {lower, upper};
}

///////////////////////////////////////////////////////////////////////////////
// FEASIBILITY CHECK
///////////////////////////////////////////////////////////////////////////////
bool ConstraintSet::checkFeasibility(const Eigen::Vector4d& state,
                                     const Eigen::Vector2d& control) const {
    // Get explicit bounds
    Eigen::Vector4d state_lower = getStateLowerBounds();
    Eigen::Vector4d state_upper = getStateUpperBounds();
    Eigen::Vector2d input_lower = getInputLowerBounds();
    Eigen::Vector2d input_upper = getInputUpperBounds();
    
    // Check state constraints [x, y, theta, delta]
    if (state(0) < state_lower(0) || state(0) > state_upper(0)) return false;  // x
    if (state(1) < state_lower(1) || state(1) > state_upper(1)) return false;  // y
    if (state(2) < state_lower(2) || state(2) > state_upper(2)) return false;  // theta
    if (state(3) < state_lower(3) || state(3) > state_upper(3)) return false;  // delta
    
    // Check input constraints [v, delta_dot]
    if (control(0) < input_lower(0) || control(0) > input_upper(0)) return false;  // v
    if (control(1) < input_lower(1) || control(1) > input_upper(1)) return false;  // delta_dot
    
    return true;
}

}  // namespace mpc_controller